lib/cfp.js

```diff
@@ -0,0 +1,325 @@
+const { getDB } = require('./firebase');
+
+const COLLECTIONS = {
+  OPTIONS: 'options',
+  BLACKLIST: 'blacklist',
+  ALIASES: 'aliases',
+  TWEETS: 'tweets',
+  HASHTAGS: 'hashtags',
+  ENTRIES: 'entries',
+  COUNTERS: 'counters',
+};
+
+const OPTIONS = {
+  SEARCH_VALUE: 'search',
+  SEARCH_HASHTAGS: 'searchHashtags',
+  LAST_RESULT_ID: 'lastResultId',
+};
+
+const extractHashtags = (text) => {
+  const regex = /(?:^|\s)(?:#([\w\d-_]+))/g;
+  const hashtags = [];
+  let match = regex.exec(text);
+  while (match) {
+    const [, hashtag] = match;
+    hashtags.push(hashtag.trim().toLowerCase());
+    match = regex.exec(text);
+  }
+
+  return hashtags;
+};
+
+const getOptions = async (db = null) => {
+  const useDB = db || getDB();
+  const result = await useDB.collection(COLLECTIONS.OPTIONS).get();
+  const options = result.docs.reduce(
+    (acc, doc) => {
+      const data = doc.data();
+      return {
+        ...acc,
+        ...{
+          [data.name]: {
+            value: data.value,
+            doc,
+          },
+        },
+      };
+    },
+    {},
+  );
+
+  options[OPTIONS.SEARCH_HASHTAGS] = {
+    value: extractHashtags(options[OPTIONS.SEARCH_VALUE].value),
+    doc: null,
+  };
+
+  return options;
+};
+
+const updateLastResultId = async (id, db = null) => {
+  const useDB = db || getDB();
+  const collection = useDB.collection(COLLECTIONS.OPTIONS);
+  const savedOption = await collection.where('name', '==', OPTIONS.LAST_RESULT_ID).get();
+  if (savedOption.empty) {
+    await collection.add({
+      name: OPTIONS.LAST_RESULT_ID,
+      value: id,
+    });
+  } else {
+    const [doc] = savedOption.docs;
+    await collection.doc(doc.id).update({ value: id });
+  }
+};
+
+const extractTweetHashtags = async (tweet, options = null) => {
+  if (tweet.entities && tweet.entities.hashtags) {
+    let useOptions;
+    if (options) {
+      useOptions = options;
+    } else {
+      useOptions = await getOptions();
+    }
```

Comment:
This block can be reduced to a single line:

```js
const useOptions = options || (await getOptions());
```

Reply:
This was my first approach; I didn't like the idea of the `await` being _hidden_ behind the or;  but I'll change it.